lib/datadog/core/telemetry/logger.rb

```diff
@@ -25,7 +25,18 @@ def error(description)
           private
 
           def instance
-            Datadog.send(:components).telemetry
+            # `allow_initialization: false` would avoid referencing the components via `safely_synchronize` (mutex)
+            # which could cause deadlock during components initialization.
+            components = Datadog.send(:components, allow_initialization: false)
```

Comment:
I still suspect the current explanation requires quite a bit of context to understand. Perhaps something like

"Component initialization uses a mutex to avoid having concurrent initialization. Trying to access the telemetry component during initialization (specifically: from the thread that's actually doing the initialization) would cause a deadlock, since accessing the components would try to recursively lock the mutex.

To work around this, we use allow_initialization: false to avoid triggering this issue.

The downside is: this leaves us unable to report telemetry during component initialization."

Something like this?